File: node_examples/game_data_generator.ts
export function generateGameplayLogs(numEntries = 500) {
  const actions = [
    "Walked to",
    "Looted",
    "Fought",
    "Crafted",
    "Talked to",
    "Completed quest",
  ];
  const locations = [
    "Forest",
    "Cave",
    "Town Square",
    "Tavern",
    "Mountain Path",
    "River Bank",
  ];
  const items = [
    "Apple",
    "Rusty Sword",
    "Leather Scraps",
    "Gold Coin",
    "Potion",
    "Arrow",
  ];
  const mobs = ["Wolf", "Bandit", "Goblin", "Rat", "Skeleton"];
  const npcs = ["Merchant", "Guard", "Villager", "Beggar"];

  const logs: string[] = [];

  // Generate noise
  for (let i = 0; i < numEntries; i++) {
    const action = actions[Math.floor(Math.random() * actions.length)];
    let log = "";
    if (action === "Walked to") {
      log = `Player walked to ${
        locations[Math.floor(Math.random() * locations.length)]
      }.`;
    } else if (action === "Looted") {
      log = `Player looted ${items[Math.floor(Math.random() * items.length)]}.`;
    } else if (action === "Fought") {
      log = `Player fought ${mobs[Math.floor(Math.random() * mobs.length)]}.`;
    } else if (action === "Talked to") {
      log = `Player talked to ${
        npcs[Math.floor(Math.random() * npcs.length)]
      }.`;
    } else {
      log = `Player performed action: ${action} in ${
        locations[Math.floor(Math.random() * locations.length)]
      }.`;
    }
    logs.push(log);
  }

  // Insert Needles
  logs.splice(
    Math.floor(numEntries / 10),
    0,
    "Player accepted quest 'Retrieve the Ancestral Hammer' from Blacksmith Gorn."
  );
  logs.splice(
    Math.floor(numEntries / 2),
    0,
    "Player was defeated by the Cave Troll but escaped."
  );
  logs.splice(
    Math.floor(numEntries / 2) + 1,
    0,
    "Player returned to Cave and defeated Cave Troll."
  );
  logs.splice(
    numEntries - 10,
    0,
    "Player found 'Ancestral Hammer' in the Troll's hoard."
  );
  logs.splice(
    numEntries - 5,
    0,
    "Player sold 'Rusty Hammer' to General Merchant."
  );

  return logs;
}

File: examples/game_data_generator.py
import random

def generate_gameplay_logs(num_entries=500):
    """Generates a list of simulated gameplay logs with hidden 'needles'."""
    
    actions = ["Walked to", "Looted", "Fought", "Crafted", "Talked to", "Completed quest"]
    locations = ["Forest", "Cave", "Town Square", "Tavern", "Mountain Path", "River Bank"]
    items = ["Apple", "Rusty Sword", "Leather Scraps", "Gold Coin", "Potion", "Arrow"]
    mobs = ["Wolf", "Bandit", "Goblin", "Rat", "Skeleton"]
    npcs = ["Merchant", "Guard", "Villager", "Beggar"]

    logs = []
    
    # Generate noise
    for _ in range(num_entries):
        action = random.choice(actions)
        if action == "Walked to":
            logs.append(f"Player walked to {random.choice(locations)}.")
        elif action == "Looted":
            logs.append(f"Player looted {random.choice(items)}.")
        elif action == "Fought":
            logs.append(f"Player fought {random.choice(mobs)}.")
        elif action == "Talked to":
            logs.append(f"Player talked to {random.choice(npcs)}.")
        else:
            logs.append(f"Player performed action: {action} in {random.choice(locations)}.")

    # Insert Needles (The important context)
    # We scatter them to make it harder for simple context windows if they get truncated or lost in noise
    
    # Needle 1: The Quest Start (Early in the logs)
    logs.insert(num_entries // 10, "Player accepted quest 'Retrieve the Ancestral Hammer' from Blacksmith Gorn.")
    
    # Needle 2: The struggle (Middle)
    logs.insert(num_entries // 2, "Player was defeated by the Cave Troll but escaped.")
    logs.insert(num_entries // 2 + 1, "Player returned to Cave and defeated Cave Troll.")
    
    # Needle 3: The Discovery (Late)
    logs.insert(num_entries - 10, "Player found 'Ancestral Hammer' in the Troll's hoard.")
    
    # Needle 4: A distraction/red herring
    logs.insert(num_entries - 5, "Player sold 'Rusty Hammer' to General Merchant.")

    return logs
